dartagnan/src/main/java/com/dat3m/dartagnan/witness/WitnessGraph.java

```diff
@@ -2,156 +2,198 @@
+    }
+
+    public Node getViolationNode() {
+        assert(nodes.stream().anyMatch(n -> n.hasAttributed(VIOLATION.toString())));
+        return nodes.stream().filter(n -> n.hasAttributed(VIOLATION.toString())).findAny().get();
+    }
+
+    public void addEdge(Edge e) {
+        nodes.add(e.getSource());
+        nodes.add(e.getTarget());
+        edges.add(e);
+    }
+
+    public Set<Node> getNodes() {
+        return nodes;
+    }
+
+    public List<Edge> getEdges() {
+        return edges;
+    }
+
+    public List<Edge> getPathToViolation() {
+        List<Edge> ret = new ArrayList<>();
+        Node cur = getViolationNode();
+        while(cur != getEntryNode()) {
+            // Local variables in lambdas have to be final, thus the copy
+            final Node copy = cur;
+            Edge nextEdge = getEdges().stream().filter(e -> e.getTarget().equals(copy)).findAny().get();
+            ret.add(nextEdge);
+            cur = nextEdge.getSource();
+        }
+        // Return path from entry to violation
+        Collections.reverse(ret);
+        return ret;
+    }
+
+    public String getProgram() {
+        return attributes.get(PROGRAMFILE.toString());
+    }
+
+    public String toXML() {
+        StringBuilder str = new StringBuilder();
+        str.append("<graph edgedefault=\"directed\">\n");
+        for (String attr : attributes.keySet()) {
+            str.append("  <data key=\"").append(attr).append("\">").append(attributes.get(attr)).append("</data>\n");
+        }
+        for (Node n : nodes) {
+            str.append(n.toXML());
+        }
+        for (Edge e : edges) {
+            str.append(e.toXML());
+        }
+        str.append("</graph>");
+        return str.toString();
+    }
+
+    private List<MemoryCoreEvent> getEventsFromEdge(Program program, Edge edge) {
+        return edge.hasCline() ?
+            program.getThreadEvents(MemoryCoreEvent.class).stream()
+                .filter(e -> e.hasMetadata(SourceLocation.class))
+                .filter(e -> e.getMetadata(SourceLocation.class).lineNumber() == edge.getCline())
+                .toList() :
+            Collections.emptyList();
+    }
+
+    public BooleanFormula encode(EncodingContext context) {
+        Program program = context.getTask().getProgram();
+        BooleanFormulaManager bmgr = context.getBooleanFormulaManager();
+        List<BooleanFormula> enc = new ArrayList<>();
+        List<MemoryCoreEvent> previous = new ArrayList<>();
+        for (Edge edge : edges.stream().filter(Edge::hasCline).toList()) {
+            List<MemoryCoreEvent> events = getEventsFromEdge(program, edge);
+            if (!previous.isEmpty() && !events.isEmpty()) {
+                enc.add(bmgr.or(Lists.cartesianProduct(previous, events).stream()
+                        .map(p -> context.edgeVariable("hb", p.get(0), p.get(1)))
+                        .toArray(BooleanFormula[]::new)));
+            }
```

Comment:
This code is kinda sketchy because it generates hb-constraints between events that cannot be hb-related since checks against the may-set are omitted. 
I understand that you need to do that to get the desired ordering between non-neighboring events but I think this should be documented somehow.